File: day1-kotlin-syntax-crash-course/java/StringTemplateExample.java
/**
 * Day 1 - String Templates in Java vs Kotlin
 *
 * Java:
 *   - String concatenation: "Hello " + name
 *   - String.format(): String.format("Hello %s", name)
 *   - Text blocks (Java 13+): """ ... """
 *   - No string templates (unlike many other languages)
 *
 * Kotlin:
 *   - String templates: "Hello $name"
 *   - Expression templates: "Length: ${name.length}"
 *   - Triple-quoted strings: """ ... """ (raw strings)
 */
public class StringTemplateExample {

    public static void main(String[] args) {

        String name = "Kotlin";
        int age = 10;
        double pi = 3.14159;
        String city = "Hanoi";

        // ═══════════════════════════════════════════════════════════════
        // 1. JAVA STRING CONCATENATION (old way)
        // ═══════════════════════════════════════════════════════════════

        // Java: concatenation with + (creates StringBuilder under the hood)
        String greeting1 = "Hello, " + name + "! You are " + age + " years old.";
        System.out.println(greeting1);

        // Java: concatenation in loops is inefficient (creates many StringBuilder objects)

        // ═══════════════════════════════════════════════════════════════
        // 2. JAVA String.format() (C-style formatting)
        // ═══════════════════════════════════════════════════════════════

        // Java: String.format with % placeholders
        String greeting2 = String.format("Hello, %s! You are %d years old.", name, age);
        System.out.println(greeting2);

        // Java: Format specifiers
        String formatted = String.format("Pi is %.2f, city is %s", pi, city);
        System.out.println(formatted);

        // Java: Various format specifiers
        System.out.println(String.format("Hex: %x, Octal: %o, Scientific: %e", 255, 255, 12345.6789));

        // ═══════════════════════════════════════════════════════════════
        // 3. JAVA TEXT BLOCKS (Java 13+)
        // ═══════════════════════════════════════════════════════════════

        // Java: Text blocks (multi-line strings)
        String json = """
                {
                    "name": "%s",
                    "age": %d
                }
                """.formatted(name, age);
        System.out.println(json);

        // ═══════════════════════════════════════════════════════════════
        // 4. JAVA: No string templates!
        // ═══════════════════════════════════════════════════════════════

        // Java has NO equivalent of Kotlin's $name or ${expr} syntax
        // The closest is String.format() or concatenation

        // ═══════════════════════════════════════════════════════════════
        // 5. COMPARISON TABLE
        // ═══════════════════════════════════════════════════════════════
        //
        //   Feature              Java 17                     Kotlin
        //   ──────────────────── ─────────────────────────── ────────────────────────
        //   Simple variable      "Hello " + name             "Hello $name"
        //   Expression           "Len: " + name.length()     "Len: ${name.length}"
        //   Multi-line           """ ... """ (Java 15+)      """ ... """
        //   Formatting           String.format("%s", name)   "Hello $name"
        //   Raw strings          No equivalent               """Hello\nWorld"""
        //   Escape characters    \n \t \\                    ${'$'} (dollar escape)
        //
        // ═══════════════════════════════════════════════════════════════

        // ═══════════════════════════════════════════════════════════════
        // 6. KEY TAKEAWAYS
        // ═══════════════════════════════════════════════════════════════
        //
        // JAVA:
        //   - Uses concatenation (+) or String.format()
        //   - Text blocks (Java 15+) for multi-line
        //   - No string templates
        //   - Format specifiers: %s, %d, %f, etc.
        //
        // KOTLIN:
        //   - $variable for simple variable interpolation
        //   - ${expression} for complex expressions
        //   - """ for raw/multi-line strings (no escaping needed)
        //   - Much more readable than Java's approach
        //
        // ═══════════════════════════════════════════════════════════════
    }
}

File: day1-kotlin-syntax-crash-course/kotlin/StringTemplateExample.kt
/**
 * Day 1 - String Templates in Kotlin
 *
 * Kotlin has built-in string template support:
 *   - $variable — simple variable interpolation
 *   - ${expression} — complex expression interpolation
 *   - """ — raw/multi-line strings (no escaping needed)
 *   - Much more readable than Java's String.format()
 */

fun main() {

    val name = "Kotlin"
    val age = 10
    val pi = 3.14159
    val city = "Hanoi"

    // ═══════════════════════════════════════════════════════════════
    // 1. SIMPLE VARIABLE INTERPOLATION ($variable)
    // ═══════════════════════════════════════════════════════════════

    val greeting1 = "Hello, $name!"
    println(greeting1)  // Hello, Kotlin!

    // Java equivalent: "Hello, " + name + "!"
    // Kotlin: Much cleaner!

    // ═══════════════════════════════════════════════════════════════
    // 2. EXPRESSION INTERPOLATION (${expression})
    // ═══════════════════════════════════════════════════════════════

    val greeting2 = "Hello, $name! You are $age years old. Next year you'll be ${age + 1}."
    println(greeting2)

    // Complex expressions
    val length = "Kotlin".length
    println("Name: $name, Length: ${name.length}")  // Can access properties directly

    // Method calls in templates
    println("Upper case: ${name.uppercase()}")
    println("First char: ${name[0]}")

    // Conditional expressions
    val status = if (age > 18) "adult" else "minor"
    println("$name is an $status")

    // ═══════════════════════════════════════════════════════════════
    // 3. RAW STRINGS (triple-quoted """)
    // ═══════════════════════════════════════════════════════════════

    // Raw strings: no escaping needed, preserve formatting
    val json = """
        {
            "name": "$name",
            "age": $age
        }
    """.trimIndent()
    println(json)

    // Java 15+ equivalent:
    // String json = """
    //         {
    //             "name": "%s",
    //             "age": %d
    //         }
    //         """.formatted(name, age);

    // Raw strings with special characters (no escaping needed!)
    val path = """C:\Users\test\file.txt"""
    println(path)

    val regex = """\d+\.\d+"""
    println(regex)

    // ═══════════════════════════════════════════════════════════════
    // 4. ESCAPING DOLLAR SIGN
    // ═══════════════════════════════════════════════════════════════

    // If you need a literal $ in the string, use ${'$'}
    val price = "Price: ${'$'}${age}"
    println(price)  // Price: $10

    // Or use raw string (no escaping needed)
    val price2 = """Price: $${age}"""
    println(price2)  // Price: $10

    // ═══════════════════════════════════════════════════════════════
    // 5. MULTI-LINE RAW STRINGS
    // ═══════════════════════════════════════════════════════════════

    val multiLine = """
        This is a
        multi-line string
        with $name interpolation
        and ${age + 1} expressions
    """.trimIndent()
    println(multiLine)

    // trimMargin() for alignment
    val aligned = """
        |Name: $name
        |Age: $age
        |City: $city
    """.trimMargin()
    println(aligned)

    // ═══════════════════════════════════════════════════════════════
    // 6. COMPARISON WITH JAVA
    // ═══════════════════════════════════════════════════════════════
    //
    // JAVA                          KOTLIN
    // ───────────────────────────── ─────────────────────────────
    // "Hello " + name               "Hello $name"
    // "Len: " + name.length()       "Len: ${name.length}"
    // String.format("Hi %s", name)  "Hi $name"
    // """
    //   line1
    //   line2
    // """                            """
    //                                  line1
    //                                  line2
    //                                """
    // No raw string equivalent       """C:\path\file"""
    //
    // ═══════════════════════════════════════════════════════════════

    // ═══════════════════════════════════════════════════════════════
    // 7. PRACTICAL EXAMPLES
    // ═══════════════════════════════════════════════════════════════

    // SQL query building
    val tableName = "users"
    val query = "SELECT * FROM $tableName WHERE age > $age"
    println(query)

    // Log message
    val userId = 123
    val action = "login"
    println("User $userId performed action: $action")

    // Template with calculation
    val items = listOf("apple", "banana", "cherry")
    println("We have ${items.size} items: ${items.joinToString(", ")}")

    // ═══════════════════════════════════════════════════════════════
    // 8. KEY TAKEAWAYS
    // ═══════════════════════════════════════════════════════════════
    //
    // 1. $name — simple variable (no braces needed for simple names)
    // 2. ${expr} — complex expressions (needed for property access, method calls)
    // 3. """ — raw strings (no escaping, preserve formatting)
    // 4. trimIndent() / trimMargin() — clean up indentation
    // 5. ${'$'} — escape dollar sign in regular strings
    // 6. Much more readable than Java's concatenation or String.format()
    //
    // ═══════════════════════════════════════════════════════════════
}
